task_processing/plugins/kubernetes/kubernetes_pod_executor.py

```diff
@@ -1,61 +1,103 @@
-
-    # TODO(TASKPROC-241): add current task state and task state history as we did for mesos
-    task_state: KubernetesTaskState = field(type=KubernetesTaskState, mandatory=True)
-    # Map of state to when that state was entered (stored as a timestamp)
-    task_state_history: PVectorType[Tuple[KubernetesTaskState, int]] = field(
-        type=PVector, factory=pvector, mandatory=True)
-
-
 class KubernetesPodExecutor(TaskExecutor):
     TASK_CONFIG_INTERFACE = KubernetesTaskConfig
 
-    def __init__(self, namespace: str) -> None:
-        self.kube_client = KubeClient()
+    def __init__(self, namespace: str, kube_config_path: Optional[str] = None) -> None:
+        self.kube_client = KubeClient(kube_config_path=kube_config_path)
         self.namespace = namespace
 
         self.task_metadata: PMap[str, KubernetesTaskMetadata] = pmap()
-        self._lock = threading.RLock()
+        self.task_metadata_lock = threading.RLock()
+
+        # we have two queues since we need to process Pod Events into something that we can place
+        # onto the queue that any application (e.g., something like Tron or Jolt) can actually use
+        # and we've opted to not do that processing in the Pod event watcher thread so as to keep
+        # that logic for the threads that operate on them as simple as possible and to make it
+        # possible to cleanly shutdown both of these.
+        # XXX: these can probably be converted to SimpleQueue once all consumers are py37+
+        self.pending_events: "Queue[Union[PodEvent, Sentinel]]" = Queue()
+        self.event_queue: "Queue[Event]" = Queue()
+
+        # TODO(TASKPROC-243): keep track of resourceVersion so that we can continue event processing
+        # from where we left off on restarts
+        self.watch = watch.Watch()
+        self.pod_event_watch_thread = threading.Thread(
+            target=self._pod_event_watch_loop,
+            # ideally this wouldn't be a daemon thread, but a watch.Watch() only checks
+            # if it should stop after receiving an event - and it's possible that we
+            # have periods with no events so instead we'll attempt to stop the watch
+            # and then join() with a small timeout to make sure that, if we shutdown
+            # with the thread alive, we did not drop any events
+            daemon=True,
+        )
+        self.pod_event_watch_thread.start()
+
+        self.pending_event_processing_thread = threading.Thread(
+            target=self._pending_event_processing_loop,
+        )
+        self.pending_event_processing_thread.start()
+
+    def _pod_event_watch_loop(self) -> None:
+        # TODO(TASKPROC-243): we'll need to correctly handle resourceVersion expiration for the case
+        # where the gap between task_proc shutting down and coming back up is long enough for data
+        # to have expired from etcd
+        for pod_event in self.watch.stream(
```

Comment:
wrap this in `while not self.stopping` and try/catch, otherwise we might be stuck with this thread dead and the rest of the system waiting for events that will never arrive.

Reply:
I wrapped both with `while not self.stopping` - we'll need to do some more work with this specific thread later to make sure that we try to resume from the correct resourceVersion - but we can do that when we start building in support for task_proc/tron restarts